src/query/api/v1/handler/prometheus/remote/match.go

```diff
@@ -103,9 +106,33 @@ func (h *PromSeriesMatchHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 		meta = meta.CombineMetadata(result.Metadata)
 	}
 
-	handleroptions.AddResponseHeaders(w, meta, opts)
+	var (
+		noopWriter = ioutil.Discard
+		renderOpts = prometheus.RenderSeriesMetadataOptions{
+			ReturnedSeriesMetadataLimit: opts.ReturnedSeriesMetadataLimit,
+		}
+	)
+	renderResult, err := prometheus.RenderSeriesMatchResultsJSON(noopWriter, results, renderOpts)
```

Comment:
Can we leave comments on these noop render calls to explain the purpose is to apply the limiting first so we can write the header before the actual body writing?

Reply:
Yeah sg.